src/core/StepRecorder.cpp

```diff
@@ -0,0 +1,465 @@
+/*
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "StepRecorder.h"
+#include "StepRecorderWidget.h"
+#include "PianoRoll.h"
+
+#include <QPainter>
+
+#include <climits>
+using std::min;
+using std::max;
+
+//for debugging: uncomment to create a win32 console, and enable DBG_PRINT
+//#define DEBUG_STEP_RECORDER
+
+#ifdef DEBUG_STEP_RECORDER
+	#define DEBUG_CREATE_WIN32_CONSOLE
+	#define DRBUG_ENABLE_PRINTS
+#endif //DEBUG_STEP_RECORDER
+
+#ifdef DEBUG_CREATE_WIN32_CONSOLE
+#include <windows.h>
+#endif
+
+#ifdef DRBUG_ENABLE_PRINTS
+#include <cstdio>
+	#define DBG_PRINT(fmt, ...) printf(fmt, ##__VA_ARGS__)
+#else
+	#define DBG_PRINT(fmt, ...) 
+#endif
+
+const int REMOVE_RELEASED_NOTE_TIME_THRESHOLD_MS = 70; 
+
+
+StepRecorder::StepRecorder(PianoRoll& pianoRoll, StepRecorderWidget& stepRecorderWidget):
+	m_pianoRoll(pianoRoll),
+	m_stepRecorderWidget(stepRecorderWidget)
+{
+ 	m_stepRecorderWidget.hide();	
+
+#if defined(DEBUG_CREATE_WIN32_CONSOLE) && defined(_WIN32)
+	//create win32 console and attach it for output 
+	if (AttachConsole(ATTACH_PARENT_PROCESS) || AllocConsole())
+	{
+		freopen("CONOUT$", "w", stdout);
+		freopen("CONOUT$", "w", stderr);
```

Comment:
Though it should not be a part of this PR, `AttachConsole` + `freopen` workaround may be useful. :+1:

Reply:
You can wrap debug prints with `LMMS_DEBUG` preprocessor definition which is defined only in debug builds. I also suggest using [Qt's functions](http://doc.qt.io/qt-5/qdebug.html) instead of plain `printf`.
For the Windows-specific console code, I'd say it should be in a separate pull request. It's totally unrelated to this feature.

Reply:
1. how do you use LMMS_DEBUG ? (where it is defined? is it triggered by some build flag?)
2. I tried to use QT's print functions before using printf and it did not work for some reason (did not output to console. I couldn't figure why)
3. I am not sure I understand, are you suggesting I should have two onging PRs on the same file?
how can see debug prints in the meanwhile? I need something I can work with and debug the component I work on... can you please provide some alternative?

Reply:
> how do you use `LMMS_DEBUG`? (where it is defined? is it triggered by some build flag?)

It's defined in the root `CMakeLists.txt`. The macro is defined in debug builds.
> I tried to use QT's print functions before using printf and it did not work

Could you let me know your OS, terminal and Qt version?
> I am not sure I understand, are you suggesting I should have two onging PRs on the same file?

I think you should move the console code to somewhere else and open a new PR.

Reply:
>It's defined in the root CMakeLists.txt. The macro is defined in debug builds.

👍

>Could you let me know your OS, terminal and Qt version?

Win10 and QT5. (is terminal question revelant to win?)

>I think you should move the console code to somewhere else and open a new PR.

I will move it, but probably won't open PR for it but rather keep it locally.